### DISCUSSION TITLE
Init module on pre-existing mentions.

### DISCUSSION THREAD

**Discussion by cavasinf** [afconsult/quill-mention#301] [Q&A] [answered]
Is there a way to init mentions on existing text ?

Like when editing an existing commentary with mentions.

**Comment by anlek** [comment]
I'm trying to do this myself, I recreated the HTML that mentions makes, however, quill now resets the curser to the beginning of the field every time you type something (and have a mention there). I was hoping there was some way to do this easier.

**Comment by cavasinf** [comment]
If denotation char is `${`.
Init Quill mention with text that contain only denotation chars will not works.
input with `{$user}` will print a text with `{$user}`.

But if I saved the result of quill mention (with HTML) 
```html
<span class="mention" data-index="0" data-denotation-char="" data-id="1" data-value="${user}">﻿
   <span contenteditable="false">
      <span class="ql-mention-denotation-char"></span>
      ${user}
   </span>
﻿</span>
```
and import it later import in a Quill editor, it works like a charm.

@anlek what kind of text/HTML do you want to import ?

**Comment by anlek** [comment]
Let's assume you have an input like: `@john, I think that @frank what to do with issue #123`

You want to parse your comment and find all words that start with `@` and `#`. I then use my search function to look up the `item`: 
```js
// See https://regexr.com/6urj4 for what this regex is doing
const mentionsRegex = /(?<=\s|^)(?<mentionChar>[@|#])(?<id>[a-zA-Z0-9]+)/g;

const parseInput = (input) => {
  let output = input; //clone input
  let match;
  while ((match = mentionsRegex.exec(output))) {
    //Extract groups that were match (group named in regex with: ?<name>)
    const { id, mentionChar } = match.groups; 

    // Using your existing lookup list, search using the id 
    // (and mentionChar if you support mutliple mentionChars)
    const item = searchYourExistingListFor(id, mentionChar);

    // If you found the item, replace the full match (like @john) with the rendered value via rendeMention(item).
    if (item) output.replace(match[0], renderMention(item));
  }

  return output;
};
```

and render it using:
```js
const renderMention = function (item, mentionChar = "#") {
  const { value, id } = item;
  // eslint-disable-next-line max-len
  return `<span class="mention" data-denotation-char="${mentionChar}" data-value="${value}" data-id="${id}">\uFEFF<span contenteditable="false"><span class="ql-mention-denotation-char">${mentionChar}</span>${value}</span>\uFEFF</span>`;
};
```

Now you can say `Quill.setContent(parseInput(myOriginalContent))`.

**Comment by anlek** [comment]
Keep in mind that I parse the mention'd quill output before saving it into the database and I replace the mention HTML content with whatever is in the `data-id` attribute.

Here is my function:
```js
const convertMentionsToIds = (input) => {
  // Create a new DOM parser
  const domParser = new DOMParser();

  // Parse input as HTML
  const content = domParser.parseFromString(input, "text/html");

  // Find all the items that have a class of `mention` (default class name of quill-mentions)
  const mentions = content.getElementsByClassName("mention");

  // clone input as final result
  let result = input;

  for (let mention of mentions) {
    // Extract know defaults
    const id = mention.getAttribute("data-id");
    const mentionChar = mention.getAttribute("data-denotation-char");

    // If you don't find an id, don't replace it.
    if (!id) continue;

    // Now, anywhere you see this mention, replace it will the mentionChar followed by ID
    result = result.replace(mention.outerHTML, `${mentionChar}${id}`);
  }

  return result;
};
```

**Comment by cavasinf** [comment]
> Keep in mind that I parse the mention'd quill output before saving it into the database and I replace the mention HTML content with whatever is in the data-id attribute.

Ok, that's what won't work with Quill Mention.
Importing text with only the mention ID will do nothing.
On my end, I saved the full HTML output to my database. By putting the HTML in the entry and running QuillJs on that entry, all HTML mentions are processed.

I gave up on making my own `ID <=> HTML Mention` JS parser to save time.
But yours is excellent.

I've give up at doing my own mention JS parser `ID <=> Mention HTML` to save me some time.
But your looks great.


> quill now resets the curser to the beginning of the field every time you type something

Is this where your problem comes from?
Do you still have the problem when you use your function without mention?

**Comment by anlek** [comment, marked as answer]
> quill now resets the curser to the beginning of the field every time you type something

This problem only happens when Quill tweaks your HTML (aka, your input isn't formatted correctly, and quill has to clean it up). This causes a re-render of the content and your curser will reset to the beginning. I currently don't have this issue because my app doesn't update quill once it's first render is done (keeping content user entered from re-rendering while they type).
